### PR TITLE
Kafka logger

### PR DESCRIPTION
**Pull request by akshayn1107**
This can go into code review phase, but there is a little bit more work left on the schemas in here which is causing a test to fail. Other then that, this is functionally correct. We might want to refactor the directories.

This is mainly a refactor from services. There might be artifacts i copied over we don't actually need (specifically with requirements.txt)

code review: @vikarmic

### CODE REVIEW

File: logger/bin/compile_pb.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+SRC_DIR="schema_proto"
```
Comment (teasherm) [inline logger/bin/compile_pb.sh]: Throw a hash bang up here to specify interpreter

File: logger/config/main.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (teasherm) [inline logger/config/main.py]: Clear this module if it's not getting imported anywhere

File: logger/nw_celery/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Background tasks for data.nerdwallet.com. These should not depend on anything inside flask_app as they could be deployed
```
Comment (teasherm) [inline logger/nw_celery/README.md]: Update for this app

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
+alembic==0.6.3
```
Comment (teasherm) [inline requirements.txt]: Prune what you can from here

**Comment by teasherm** [comment]
Off to a good start IMO, we'll see what @vikarmic has to say

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,27 @@
+celery==3.1.10
```
Comment (akshayn1107) [inline requirements.txt]: I've begun pruning, but not sure if I got everything (in fact I'm sure I didn't). Thoughts on how to proceed?

Comment (teasherm) [inline requirements.txt]: Sure way is to start with fresh venv, grep the codebase for package imports, pip install all of them then pip freeze

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,15 @@
+default: test
```
Comment (vikarmic) [inline Makefile]: Fuck yeah, makefiles.

File: logger/app/__init__.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+from flask import jsonify
+
+from .. import factory
+from ..core.errors import NWError
+from ..helpers import JSONEncoder
+
+
+def create_app(settings_override=None):
+    """Returns the NWAdmin API application instance"""
```
Comment (vikarmic) [inline logger/app/__init__.py]: Nitpick: no longer the NWAdmin app.

File: logger/app/endpoints.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+from logger.core.loggers import InvalidPayloadException
+import wsgi
+
+bp = Blueprint('logger', __name__)
+
+@route(bp, '/<log_engine>', methods=['POST'])
+def log_message(log_engine):
+    """
+    Generic logger handler: log a message with a registered logEngine
+    Loggers are loaded semidynamically; see nerdlogger.loggers for more details.
```
Comment (vikarmic) [inline logger/app/endpoints.py]: Make sure this points to the right file.

File: logger/app/endpoints.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    We use Celery to run the loggers asynchronously, so the celery worker must be running when
+      you log a message (use the "run_celery" script in src)
+    Will return "OK" on success, and "error: [error msg]" on failure.
+    """
```
Comment (vikarmic) [inline logger/app/endpoints.py]: My initial error handling here was a bit primitive; do we want to spend some time and clean it up, or do we think this is sufficient?

File: logger/core/loggers.py [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+
+class KafkaLogger(AbstractLogger):
+    """
+    Logger which sends the data over to our kafka cluster to be logged.
+    """
+
+    __logger__ = "kafka"
+
+    # TODO: should we move these constants to a more central file if
+    # we will be using them elsewhere?
```
Comment (vikarmic) [inline logger/core/loggers.py]: Might be worth dropping them in config/environment.php.

File: logger/core/loggers.py [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+
+    def get_and_set(self, i):
+        with self.locks[i]:
+            if self.array[i]:
+                error_str = "Received request {0} twice".format(i)
+                raise InvalidPayloadException(error_str)
+            self.array[i] = True
+            return
+
+class LoadTestLogger(AbstractLogger):
```
Comment (vikarmic) [inline logger/core/loggers.py]: I recommend moving your testing code out to a separate file, for separation of concerns.

File: requirements.txt (hunk 5 above)
Comment (vikarmic) [inline requirements.txt]: Of this, my suspicion is that anything involving SQL, mock, pygeoip and pycrypto can be deleted; possibly also MarkupSafe but I'm not as sure there.  Tom's suggestion is probably the better call.

**Comment by vikarmic** [comment]
:+1:

**Pull request merged**